File: src/javascript/algorithms/doublyLinkedlist/3.swapNodePairs/swapNodePairs.js
class Node {
    constructor(value) {
        this.value = value;
        this.next = null;
        this.prev = null;
    }
}

class DoublyLinkedList {

    constructor(value) {
        this.head = null;
        this.tail = null;
        this.length = 0;
    }

    push(value) {
        const newNode = new Node(value);
        if(!this.head) {
            this.head = newNode;
            this.tail = newNode;
        } else {
            this.tail.next = newNode;
            newNode.prev = this.tail;
            this.tail = newNode;
        }
        this.length++;
        return this;
    }

    swapPairs() {
        if(this.length < 2) return;
        let current = this.head;
        let newHead = this.head.next;
        
        while(current !== null && current.next !== null) {
            let firstNext = current.next;
            let secondNext = current.next.next;
            let prev = current.prev;
            
            if(prev) prev.next = firstNext;
            firstNext.next = current;
            current.next = secondNext;
            
            if(secondNext) secondNext.prev = current;
            current.prev = firstNext;
            firstNext.prev = prev;
            
            current = current.next;
        }
        this.head = newHead;
    }

}

let myDoublyLinkedList1 = new DoublyLinkedList();
myDoublyLinkedList1.push(1);
myDoublyLinkedList1.push(2);
myDoublyLinkedList1.push(3);
myDoublyLinkedList1.push(4);
myDoublyLinkedList1.push(5);
myDoublyLinkedList1.push(6);

let myDoublyLinkedList2 = new DoublyLinkedList();
myDoublyLinkedList2.push(1);
myDoublyLinkedList2.push(2);
myDoublyLinkedList2.push(3);
myDoublyLinkedList2.push(4);
myDoublyLinkedList2.push(5);

console.log("Before swap node pairs")
console.dir(myDoublyLinkedList1, {depth: null});
console.log("After swap node pairs")
myDoublyLinkedList1.swapPairs();
console.dir(myDoublyLinkedList1, {depth: null});

console.log("Before swap node pairs")
console.dir(myDoublyLinkedList2, {depth: null});
console.log("After swap node pairs")
myDoublyLinkedList2.swapPairs();
console.dir(myDoublyLinkedList2, {depth: null});

File: src/python/algorithms/doublyLinkedlist/3.swapNodePairs/swap_node_pairs.py
"""
Swap Node Pairs in Doubly Linked List

Swap every two adjacent nodes in a doubly linked list.

Time Complexity: O(n)
Space Complexity: O(1)
"""


class Node:
    def __init__(self, value):
        self.value = value
        self.next = None
        self.prev = None


class DoublyLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
        self.length = 0
    
    def push(self, value):
        """Add a node to the end of the list."""
        new_node = Node(value)
        if not self.head:
            self.head = new_node
            self.tail = new_node
        else:
            self.tail.next = new_node
            new_node.prev = self.tail
            self.tail = new_node
        self.length += 1
        return self
    
    def swap_pairs(self):
        """Swap every two adjacent nodes. TC: O(n), SC: O(1)"""
        if self.length < 2:
            return
        
        current = self.head
        new_head = self.head.next
        
        while current and current.next:
            first_next = current.next
            second_next = current.next.next
            prev = current.prev
            
            if prev:
                prev.next = first_next
            first_next.next = current
            current.next = second_next
            
            if second_next:
                second_next.prev = current
            current.prev = first_next
            first_next.prev = prev
            
            # Update tail if needed
            if not current.next:
                self.tail = current
            
            current = current.next
        
        self.head = new_head
    
    def print_list(self):
        """Print the list values."""
        values = []
        curr = self.head
        while curr:
            values.append(str(curr.value))
            curr = curr.next
        return " <-> ".join(values)


# Test cases
if __name__ == "__main__":
    dll = DoublyLinkedList()
    dll.push(1)
    dll.push(2)
    dll.push(3)
    dll.push(4)
    
    print(f"Before swap pairs: {dll.print_list()}")
    dll.swap_pairs()
    print(f"After swap pairs: {dll.print_list()}")
    # 1 <-> 2 <-> 3 <-> 4 becomes 2 <-> 1 <-> 4 <-> 3
    
    dll2 = DoublyLinkedList()
    dll2.push(1)
    dll2.push(2)
    dll2.push(3)
    dll2.push(4)
    dll2.push(5)
    
    print(f"Before swap pairs: {dll2.print_list()}")
    dll2.swap_pairs()
    print(f"After swap pairs: {dll2.print_list()}")
    # 1 <-> 2 <-> 3 <-> 4 <-> 5 becomes 2 <-> 1 <-> 4 <-> 3 <-> 5
